backend-lib/src/main/rust/lib.rs

```diff
@@ -438,17 +421,128 @@ pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_createAcc
+///   `recover_until` height, exclusive.
+/// - `key_source`: A string identifier or other metadata describing the source of the seed.
+///   This is treated as opaque metadata by the wallet backend; it is provided for use by
+///   applications which need to track additional identifying information for an account.
+///
+/// # Panics
+///
+/// Panics if the length of the seed is not between 32 and 252 bytes inclusive.
+#[unsafe(no_mangle)]
+pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_importAccountUfvk<'local>(
+    mut env: JNIEnv<'local>,
+    _: JClass<'local>,
+    db_data: JString<'local>,
+    network_id: jint,
+    account_name: JString<'local>,
+    key_source: JString<'local>,
+    ufvk_str: JString<'local>,
+    treestate: JByteArray<'local>,
+    recover_until: jlong,
+    purpose: jint,
+    seed_fingerprint_bytes: JByteArray<'local>,
+    hd_account_index_raw: u32,
+) -> jobject {
+    let res = catch_unwind(&mut env, |env| {
+        let network = parse_network(network_id as u32)?;
+        let mut db_data = wallet_db(env, network, db_data)?;
+        let ufvk_str = java_string_to_rust(env, &ufvk_str);
+        let ufvk = UnifiedFullViewingKey::decode(&network, &ufvk_str).map_err(|e| {
+            anyhow!(
+                "Value \"{}\" did not decode as a valid UFVK: {}",
+                ufvk_str,
+                e
+            )
+        })?;
+        let treestate = TreeState::decode(&env.convert_byte_array(treestate).unwrap()[..])
+            .map_err(|e| anyhow!("Invalid TreeState: {}", e))?;
+        let recover_until = recover_until.try_into().ok();
+
+        let birthday =
+            AccountBirthday::from_treestate(treestate, recover_until).map_err(|e| match e {
+                BirthdayError::HeightInvalid(e) => {
+                    anyhow!("Invalid TreeState: Invalid height: {}", e)
+                }
+                BirthdayError::Decode(e) => {
+                    anyhow!("Invalid TreeState: Invalid frontier encoding: {}", e)
+                }
+            })?;
+
+        let account_name = java_string_to_rust(env, &account_name);
+        let key_source = java_nullable_string_to_rust(env, &key_source);
+
+        let seed_fingerprint =
+            <[u8; 32]>::try_from(&env.convert_byte_array(seed_fingerprint_bytes)?[..])
+                .ok()
+                .map(SeedFingerprint::from_bytes);
+        let hd_account_index = zip32::AccountId::try_from(hd_account_index_raw).ok();
+
+        let derivation = seed_fingerprint
+            .zip(hd_account_index)
+            .map(|(seed_fp, idx)| Zip32Derivation::new(seed_fp, idx));
+
+        let purpose = match purpose {
+            0 => Ok(AccountPurpose::Spending { derivation }),
+            1 => Ok(AccountPurpose::ViewOnly),
+            _ => Err(anyhow!(
+                "Account purpose must be either 0 (Spending) or 1 (ViewOnly)"
+            )),
+        }?;
```

Comment:
`seed_fingerprint`, `hd_account_index`, and `derivation` are only used if `purpose` is 0. Is the following suggestion equivalent, or are we relying on deriving those values for error checking?
```suggestion
        let purpose = match purpose {
            0 => {
                let seed_fingerprint =
                    <[u8; 32]>::try_from(&env.convert_byte_array(seed_fingerprint_bytes)?[..])
                    .ok()
                    .map(SeedFingerprint::from_bytes);
                let hd_account_index = zip32::AccountId::try_from(hd_account_index_raw).ok();

                let derivation = seed_fingerprint
                    .zip(hd_account_index)
                    .map(|(seed_fp, idx)| Zip32Derivation::new(seed_fp, idx));

                Ok(AccountPurpose::Spending { derivation })
            },
            1 => Ok(AccountPurpose::ViewOnly),
            _ => Err(anyhow!(
                "Account purpose must be either 0 (Spending) or 1 (ViewOnly)"
            )),
        }?;
```

Reply:
Fixed in #1656